core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -166,6 +168,29 @@ public Optional<Rel> visit(Cross cross) throws RuntimeException {
             .build());
   }
 
+  @Override
+  public Optional<Rel> visit(HashJoin hashJoin) throws RuntimeException {
+    var left = hashJoin.getLeft().accept(this);
+    var right = hashJoin.getRight().accept(this);
+    var leftKeys = hashJoin.getLeftKeys();
+    var rightKeys = hashJoin.getRightKeys();
+    var postFilter = hashJoin.getPostJoinFilter().flatMap(t -> visitExpression(t));
+    if (allEmpty(left, right, leftKeys, rightKeys, postFilter)) {
+      return Optional.empty();
+    }
```

Comment:
The purpose of the RelCopyOnWriteVisitor is to produce a copy of a plan tree ONLY if some part of the tree has changed.

Here, `left`, `right` and `postFilter` are all empty if they have not been changed by calling `accept` or `visitExpression`, so it makes sense to include them in the `allEmpty` check 

`leftKeys` and `rightKeys` just return optional if they are not set (possibly another reason to have them be required in HashJoin). As is, they shouldn't be in the allEmpty check.

There is a deeper issue in that field references possibly _should_ be recomputed as part of the RelCopyOnWriteVisitor (see https://github.com/substrait-io/substrait-java/issues/185), but I think that's outside the scope of the PR. Noting it here so it gets linked to the issue.

Reply:
Shall we create an issue? And make sure this is addressed.

Reply:
I think https://github.com/substrait-io/substrait-java/issues/185 should be enough. It captures the overall issue.